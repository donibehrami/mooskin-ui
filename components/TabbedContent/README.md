# MooSkin UI - TabbedContent Component

The MooSkin TabbedContent Component is like a tabbed window, with as many tabs as you like.

___

### Usage

To start using the TabbedContent Component first you have to Import its two main components

```
Import {TabbedContent, Tab, Content, Header} from 'mooskin';
```
or
```
import TabbedContent, {Tab, Content, Header} from 'mooskin/lib/TabbedContent';
```

And then you can simply start using it by typing

```
<TabbedContent>
    <Tab>
        <Header>Moosend</Header>
        <Content>Table</Content>
    </Tab>
    <Tab active>
        <Header>Home</Header>
        <Content>Some Content</Content>
    </Tab>
    <Tab>
        <Header>About</Header>
        <Content>Blla blla</Content>
    </Tab>
</TabbedContent>
```

Many props can be passed to individual components which will render them differently depending on the passed props. This can be from custom classes or styles, to more advanced props like header aligning or other types of TabbedContent.

### Examples


```
<TabbedContent>
    <Tab>
        <Header>Moosend</Header>
        <Content>Table</Content>
    </Tab>
    <Tab active>
        <Header>Home</Header>
        <Content>Some Content</Content>
    </Tab>
    <Tab>
        <Header>About</Header>
        <Content>Blla blla</Content>
    </Tab>
</TabbedContent>
```

Can be used without importing the Tab component with shorthand dot(.) notation.

```
Import {TabbedContent} from 'mooskin';

<TabbedContent>
    <TabbedContent.Tab>
        <TabbedContent.Header>Moosend</Header>
        <TabbedContent.Content>Table</Content>
    </TabbedContent.Tab>
    <TabbedContent.Tab active>
        <TabbedContent.Header>Home</Header>
        <TabbedContent.Content>Some Content</Content>
    </TabbedContent.Tab>
    <TabbedContent.Tab>
        <TabbedContent.Header>About</Header>
        <TabbedContent.Content>Blla blla</Content>
    </TabbedContent.Tab>>
</TabbedContent.TabbedContent>
```

The type component will alter the styles of the TabbedContent, like a radio styled component. Also if u want headers to appear vertically with the content on the right, there is the `vertical` prop.

```
<TabbedContent type="radio" vertical >
    <Tab>
        <Header>Moosend</Header>
        <Content>Table</Content>
    </Tab>
    <Tab active>
        <Header>Home</Header>
        <Content>Some Content</Content>
    </Tab>
    <Tab>
        <Header>About</Header>
        <Content>Blla blla</Content>
    </Tab>
</TabbedContent>
```

<div class="playground-doc">

## Supported attributes for TabbedContent

* `id` - id of the element
* `vertical` - tab headers will be aligned vertically, content on the right
* `type` - renders a different tabbed content based on the passed value, for now only `radio` is supported.
* `alignHeaders` - align tab headers to the left/center/right or top/center/bottom for vertical
* `className` - css class
* `style` - extra styles for the main container

## Supported attributes for Tab 

* `active` - sets which tab is active, the first is active by default
* `title` - (required) sets the title for the tab
* `className` - css class
* `style` -  extra styles 

## Supported attributes for Header 

* `width` - width of the header, mostly used when alignHeaders is used
* `className` - css class
* `style` -  extra styles 
* `onClick` - onClick callback function when the header is clicked

## Supported attributes for Content 

* `className` - css class
* `style` -  extra styles 

</div>

Allthough these attributes are supported, most of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)