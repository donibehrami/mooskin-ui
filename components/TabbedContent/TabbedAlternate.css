@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.containerH {
    display: flex;
    flex-direction: column;
}

.containerV {
    display:flex;
    flex-direction: row;
}

.headingV {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    flex-shrink: 1; 
    border-radius: 8px;
    background: #fff !important;
    /* border-right: 1px solid var(--moosendBlue); */
}

.headingH {
    display: flex;
    flex-grow: 1;
    flex-shrink: 1;
    flex-direction: row;
    border-radius: 8px;
    background: #fff !important;
    /* border-bottom: 1px solid var(--moosendBlue); */
}

.content{
    color: #7b7b7b;
    font-size: 0.875rem;
}

.contentH {
    flex-grow: 1;
    flex-shrink: 1;
}

.contentV {
    flex: 2;
}

.header {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    border-radius: 0px;
    /* color: var(--moosendBlue); */
    color: #aaa;
    font-size: 1rem;
    background: transparent;
    transition: background-color 0.3s;
    /* border: 1px solid var(--moosendBlue); */
    padding: 13px;
    cursor: pointer;
    user-select: none;
}

.header label {
    display: flex;
    justify-content: center;
    flex-direction: column;
    cursor: pointer;
    font-size: 0.875rem;
    width: 100%;
    height: 100%;
}

.activeHeader {
    background: var(--moosendBlue);
    color: #fff;
    margin: 0;
}

.activeHeader:first-child {
    border-bottom-left-radius: 8px;
    border-top-left-radius: 8px;
}

.activeHeader:last-child {
    border-bottom-right-radius: 8px;
    border-top-right-radius: 8px;
}

.activeHeader:hover {
    background: var(--moosendBlue);
    border-color: var(--moosendBlue);
}

.visible {
    display: block;
}

.invisible {
    display: none;
}

@media (--xSmall){

    .headingV{
        margin: 0;
        padding: 0;
        border-right: none;
        margin-bottom: 10px;
        padding-bottom: 10px;
        border-bottom: 1px solid var(--moosendBlue);
    }

    .headingH{
        flex-direction: column;
    }

    .containerV{flex-direction: column}

} 
