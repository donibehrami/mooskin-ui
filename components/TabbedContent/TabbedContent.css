@import url('https://fonts.googleapis.com/css?family=Convergence');
@import url('https://fonts.googleapis.com/icon?family=Material+Icons');
@import '../_utils/globals/variables.css';
@import '../_utils/globals/screens.css';

.container{
    display: flex;
    width: 100%;
}

.containerV {
    flex-direction: row;
}

.containerH {
    flex-direction: column;
}

.heading {
    display:flex;
    flex-direction: row;
}

.headingH {
    flex: 1;
    border-bottom:10px solid #efefef;
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
}

.headingV {
    flex-direction: column;
    flex: 1;
    border-right:10px solid #efefef;
    border-top-left-radius: 5px;
    border-bottom-left-radius: 5px;
}

.contentH {
    flex: 1;
}

.contentV {
    flex: 2;
}

.header {
    padding: 30px 0;
    display: inline-block;
    flex-grow:1;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
    transition: all 400ms ease;
    border-top-left-radius: inherit;
    border-top-right-radius: inherit;
    border-bottom-left-radius: inherit;
}

.header span{
    color: #575757;
    font-family: Convergence,sans-serif;
    font-size: 1.125rem;
    /*font-size: 18px;*/
}

.tab{
    display: block;
}

.activeHeader {
    background: var(--inputBackground);
}

.inactiveHeader {
    background: #fff;
}

.inactiveHeader:hover {
    background: rgba(245, 245, 245, 0.5);
}

.visible {
    display: block;
}

.invisible {
    display: none;
}

@media (--deviceSmall){

    .heading{
        height: 10vw;
    }

    .heading span{
        font-size: 4vw;
    }

}

@media (--xSmall){
    .containerV{flex-direction: column}

    .headingV{
        flex-direction: row;
        border-bottom:10px solid #efefef;
        border-right: none;
    }

}
/* 
@media (--small){
    .containerV{flex-direction: column}

    .headingV{
        flex-direction: row;
        border-bottom:10px solid #efefef;
        border-right: none;
    }
} */

@media (--deviceXsmall){

    .heading span{
        font-size: 3.5vw;
    }

    .containerH {
        font-size: 3vw;
    }

    .contentV {
        font-size: 3vw;
    }

}

@media (--deviceSmall){

    .heading span{
        font-size: 3.5vw;
    }

    .containerH {
        font-size: 3vw;
    }

    .contentV {
        font-size: 3vw;
    }

}