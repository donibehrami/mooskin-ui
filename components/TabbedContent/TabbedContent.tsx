import * as React from 'react';

import normal from './TabbedContent.css';
import radio from './TabbedRadio.css';

export interface ITabbedContentProps {

    /** id of the component */
    id?: string;

    /** type of the tabbed content normal/radio/advanced */
    type?: 'normal' | 'radio';

    /** align headers left/center/right */
    alignHeaders?: string;

    /** vertical styles of the tabbed content */
    vertical?: boolean;

    /** container class */
    className?: string;

    /** override styles */
    style?: React.CSSProperties;

    /** children here can only be Tab elements */
    children?: React.ReactElement<ITabProps> | Array<React.ReactElement<ITabProps>>;
}

export interface ITabProps {

    /** if active */
    active?: boolean;

    /** container class */
    className?: string;

    /** override styles */
    style?: React.CSSProperties;

    /** children here can only be Tab elements */
    children?: Array<React.ReactElement<IHeaderProps> | React.ReactElement<IContentProps>>;
}

export interface IHeaderProps {

    /** if active (inherited from parent) */
    active?: boolean;

    /** type of tabbed content should reflect on the header (inherited from parent) */
    type?: string;

    /** how headers should be aligned (inherited from parent) */
    align?: string;

    /** wether the headers should be aligned vertically (inherited from parent) */
    vertical?: boolean;

    /** define header width, mostly used with alignHeaders */
    width?: number;

    /** container class */
    className?: string;

    /** override styles */
    style?: React.CSSProperties;

    /** children here can only be Tab elements */
    children?: any;

    /** callback function when a header is clicked */
    onClick?: (e: React.MouseEvent<HTMLElement>) => void;
}

export interface IContentProps {

    /** if active (inherited from parent) */
    active?: boolean;

    /** container class */
    className?: string;

    /** override styles */
    style?: React.CSSProperties;

    /** children here can only be Tab elements */
    children?: any;
}

export interface ITabbedContentState {
    activeTab?: number;
}

export default class TabbedContent extends React.Component<ITabbedContentProps, ITabbedContentState>{

    static defaultProps = {
        activeTab: 0,
        className: '',
        style: {},
        type: 'normal'
    };

    static Tab: React.StatelessComponent<ITabProps>;
    static Header: React.StatelessComponent<IHeaderProps>;
    static Content: React.StatelessComponent<IContentProps>;

    constructor(props: ITabbedContentProps){
        super(props);

        this.state = {
            activeTab: this.getActiveTab()
        };
    }

    render() {

        const style = this.getClasses();

        const containerStyles = !this.props.vertical ? style.containerH : style.containerV;
        const headingStyles = !this.props.vertical ? style.headingH : style.headingV;
        const contentStyles = !this.props.vertical ? style.contentH : style.contentV;

        const verticalAlign = this.props.alignHeaders && this.props.vertical ? style.headingVAlign : '';
        const align = this.getAlign(style);

        const {headers, contents} = this.riteOfRakshir();

        return(
            <div id={this.props.id} className={`tabbed-content-component ${style.container} ${containerStyles}`}>
                <div className={`${style.heading} ${headingStyles} ${align} ${verticalAlign}`}>
                    {headers}
                </div>
                <div className={`${style.content} ${contentStyles}`}>
                    {contents}
                </div>
            </div>
        );
    }

    riteOfRakshir = () => {

        const headers: Array<React.ReactElement<IHeaderProps>> = [];
        const contents: Array<React.ReactElement<ITabProps>> = [];

        React.Children.forEach(this.props.children, (child, index) => {

            if (React.isValidElement<ITabProps>(child)){
                const {header, content} = this.assignPropsToChildren(child.props.children, index);

                headers.push(header);

                contents.push(content);
            }

        });

        return {headers, contents};
    }

    assignPropsToChildren = (children: any, index: number) => {

        const headerProps: Partial<IHeaderProps & {key: number}> = {
            active: this.state.activeTab === index,
            align: this.props.alignHeaders,
            key: index,
            onClick: this.onClickHeader(index, children[0]),
            type: this.props.type,
            vertical: this.props.vertical,
        };

        const contentProps: Partial<IContentProps & {key: number}> = {
            active: this.state.activeTab === index,
            key: index
        };

        const header: any = React.cloneElement(children[0], headerProps);
        const content: any = React.cloneElement(children[1], contentProps);

        return {header, content};
    }

    onClickHeader = (tabIndex: number, header: React.ReactElement<IHeaderProps>) => {
        return (e: React.MouseEvent<HTMLElement>) => {
            this.setState({activeTab: tabIndex});
            header.props.onClick && header.props.onClick(e);
        };
    }

    getActiveTab(): number {
        let activeTab = 0;
        const childrenArray = React.Children.toArray(this.props.children);

        for (const [index, value] of childrenArray.entries()){
            if (React.isValidElement<ITabProps>(value) && value.props.active){
                activeTab = index;
            }
        }

        return activeTab;
    }

    getClasses = () => {
        switch (this.props.type) {
            case 'normal':

                return normal;

            case 'radio':

                return radio;

            default:

                return normal;
        }
    }

    getAlign = (style: any) => {
        if (this.props.alignHeaders === 'right' || this.props.alignHeaders === 'bottom'){
            return !this.props.vertical ? style.headingHRight : style.headingVRight;
        } else if (this.props.alignHeaders === 'left' || this.props.alignHeaders === 'top'){
            return '';
        } else {
            return style.headingCenter;
        }
    }
}

export const Tab: React.StatelessComponent<ITabProps> = (props) => {

    return(
        <div className={`tab-component ${props.className}`} style={props.style}>
            {props.children}
        </div>
    );
};

Tab.defaultProps = {
    className: '',
    style: {}
};

export const Header: React.StatelessComponent<IHeaderProps> = (props) => {

    const getClasses = () => {
        switch (props.type) {
            case 'normal':

                return normal;

            case 'radio':

                return radio;

            default:

                return normal;
        }
    };

    const style = getClasses();

    const activeTab = props.active ? style.activeHeader : style.inactiveHeader;

    const input = props.type === 'radio' ? (
        <input
            type="radio"
            checked={props.active}
            readOnly
        />
    ) : '';

    const align = props.align ? style.headerAlign : '';

    return(
        <div
            className={`tab-header ${style.header} ${activeTab} ${align} ${props.className}`}
            style={{width: props.width, ...props.style}}
            onClick={props.onClick}
        >
            {input}
            {props.children}
        </div>
    );
};

Header.defaultProps = {
    className: '',
    style: {}
};

export const Content: React.StatelessComponent<IContentProps> = (props) => {

    const displayClass = !props.active ? normal.invisible : normal.visible;

    return(
        <div className={`content-component ${displayClass} ${props.className}`} style={props.style}>
            {props.children}
        </div>
    );
};

Content.defaultProps = {
    className: '',
    style: {}
};
