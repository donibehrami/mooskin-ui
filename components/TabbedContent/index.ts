import TabbedContent, {Tab} from './TabbedContent';

TabbedContent.Tab = Tab;

export default TabbedContent;
export {Tab};
