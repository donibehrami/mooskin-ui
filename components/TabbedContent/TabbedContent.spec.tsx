import * as React from 'react';
import TabbedContent, {Header, Tab} from './TabbedContent';

import { mount, render, shallow } from 'enzyme';

describe('TabbedContent', () => {

    test('renders TabbedContent properly according to snapshot', () => {

        const component = shallow(
            <TabbedContent id="5" className="mySwitch" style={{color: 'black'}}>
                <Tab title="title1" iconClass="iconclass">asdasd1</Tab>
                <Tab title="title2" materialIcon="home" active>asdasd2</Tab>
                <Tab title="title3" className="classname">asdasd3</Tab>
            </TabbedContent>
        );

        expect(component.find('Header').length).toBe(3);
        expect(component.find('Tab').length).toBe(3);

        expect(component.find('Tab').first().prop('children')).toBe('asdasd1');

        expect(component).toMatchSnapshot();
    });

    test('renders Tab properly according to snapshot', () => {
        const component = shallow(
        <Tab title="title1" iconClass="iconclass" materialIcon="home" style={{color: 'blue'}} active>asdasd1</Tab>
        );

        expect(component).toMatchSnapshot();
    });

    test('renders Header properly according to snapshot', () => {
        const component = shallow(
            <Header
                title="title1"
                iconClass="iconclass"
                materialIcon="home"
                onClick={() => null}
                active
            />
        );

        expect(component).toMatchSnapshot();
    });

    test('renders TabbedContent properly with icon class and material icon', () => {

        const component = mount(
            <TabbedContent>
                <Tab title="title1" iconClass="iconclass">asdasd1</Tab>
                <Tab title="title2" materialIcon="home">asdasd2</Tab>
            </TabbedContent>
        );

        expect(component.find('.tab-header i.iconclass').length).toBe(1);
        expect(component.find('.tab-header i.material-icons').text()).toBe('home');

    });

    test('renders without any <i> elements when no iconClass or materialIcon is provided', () => {

        const component = mount(
            <TabbedContent>
                <Tab title="title1">asdasd1</Tab>
                <Tab title="title2">asdasd2</Tab>
            </TabbedContent>
        );

        expect(component.find('.tab-header i.iconclass').length).toBe(0);
        expect(component.find('.tab-header i.material-icons').length).toBe(0);

    });

    test('changes the state and displays appropriate tab when one of the headers is clicked and back', () => {

        const component = mount(
            <TabbedContent>
                <Tab title="title1">asdasd1</Tab>
                <Tab title="title2" active>asdasd2</Tab>
            </TabbedContent>
        );

        expect(component.state('activeTab')).toBe(1);
        expect(component.find('.tab-header').first().hasClass('inactiveHeader')).toBeTruthy();
        expect(component.find('.tab-header').last().hasClass('activeHeader')).toBeTruthy();
        expect(component.find('.tab-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.tab-content').last().hasClass('visible')).toBeTruthy();

        component.find('.tab-header').first().simulate('click');

        expect(component.state('activeTab')).toBe(0);
        expect(component.find('.tab-header').last().hasClass('inactiveHeader')).toBeTruthy();
        expect(component.find('.tab-header').first().hasClass('activeHeader')).toBeTruthy();
        expect(component.find('.tab-content').last().hasClass('invisible')).toBeTruthy();
        expect(component.find('.tab-content').first().hasClass('visible')).toBeTruthy();

        component.find('.tab-header').last().simulate('click');

        expect(component.state('activeTab')).toBe(1);
        expect(component.find('.tab-header').first().hasClass('inactiveHeader')).toBeTruthy();
        expect(component.find('.tab-header').last().hasClass('activeHeader')).toBeTruthy();
        expect(component.find('.tab-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.tab-content').last().hasClass('visible')).toBeTruthy();

    });
});
