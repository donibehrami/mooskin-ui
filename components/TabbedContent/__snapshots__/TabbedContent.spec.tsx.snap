// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabbedContent renders TabbedContent properly according to snapshot 1`] = `
<div
  class="tabbed-content-component container"
>
  <div
    class="heading"
  >
    <div
      class="tab-header header inactiveHeader"
    >
      <i
        class="header-icon iconclass"
      />
      <span>
        title1
      </span>
    </div>
    <div
      class="tab-header header activeHeader"
    >
      <i
        class="header-icon material-icons"
      >
        home
      </i>
      <span>
        title2
      </span>
    </div>
    <div
      class="tab-header header inactiveHeader"
    >
      <span>
        title3
      </span>
    </div>
  </div>
  <div
    class="body"
  >
    <div
      class="tab-content tab invisible"
    >
      asdasd1
    </div>
    <div
      class="tab-content tab visible"
    >
      asdasd2
    </div>
    <div
      class="tab-content tab invisible"
    >
      asdasd3
    </div>
  </div>
</div>
`;
